### PR TITLE
fixed #187

### LINKED ISSUE #187: Bug deploying rest services - maybe missing <hub:plugins> config ?
I copied an old REST directory into my conformance directory of an active hub (the plugins area). I edited the files (one transform and one extension) in order to ensure they were deployed.

Deploy failed with ML stack trace:

2016-04-07 13:15:15.407 Info: data-hub-STAGING: <error:error xsi:schemaLocation="http://marklogic.com/xdmp/error error.xsd" xmlns:error="http://marklogic.com/xdmp/error" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
2016-04-07 13:15:15.407 Info: data-hub-STAGING:   <error:code>XDMP-AS</error:code>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:   <error:name>err:XPTY0004</error:name>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:   <error:xquery-version>1.0-ml</error:xquery-version>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:   <error:message>Invalid coercion</error:message>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:   <error:format-string>XDMP-AS: (err:XPTY0004) $uri as xs:string -- Invalid coercion: () as xs:string</error:format-string>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:   <error:retryable>false</error:retryable>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:   <error:expr>$uri as xs:string</error:expr>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:   <error:data>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:     <error:datum>()</error:datum>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:     <error:datum>xs:string</error:datum>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:   </error:data>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:   <error:stack>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:     <error:frame>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:       <error:uri>/com.marklogic.hub/lib/flow-lib.xqy</error:uri>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:       <error:line>623</error:line>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:       <error:column>26</error:column>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:       <error:operation>flow:validate-entities()</error:operation>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:       <error:variables>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:         <error:variable>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:           <error:name xmlns="http://marklogic.com/data-hub/flow-lib">errors</error:name>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:           <error:value>json:array()</error:value>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:         </error:variable>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:         <error:variable>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:           <error:name xmlns="http://marklogic.com/data-hub/flow-lib">options</error:name>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:           <error:value>map:map()</error:value>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:         </error:variable>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:         <error:variable>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:           <error:name xmlns="http://marklogic.com/data-hub/flow-lib">entity</error:name>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:           <error:value>&lt;entity xmlns="http://marklogic.com/data-hub"&gt;&lt;name&gt;consumer&lt;/name&gt;&lt;flows&gt;&lt;flow&gt;&lt;name&gt;REST&lt;/name&gt;&lt;entity&gt;consu...&lt;/entity&gt;</error:value>
2016-04-07 13:15:15.407 Info: data-hub-STAGING:         </error:variable>

Stack is flow-lib:623 which is the last line in this chunk:
    (: validate plugins :)
    let $_ :=
      for $plugin in $flow/hub:plugins/hub:plugin
      let $destination := $plugin/@dest
      let $module-uri := $plugin/@module
      let $module-name := hul:get-module-name($module-uri)  <---- error here

and QS stack:

13:15:15.418 [http-nio-8080-exec-5] ERROR o.a.c.c.C.[.[.[.[dispatcherServlet] - Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is com.marklogic.client.FailedRequestException: Local message: failed to read resource at resources/validate: Internal Server Error. Server Message: XDMP-AS: (err:XPTY0004) $uri as xs:string -- Invalid coercion: () as xs:string . See the MarkLogic server error log for further detail.] with root cause
com.marklogic.client.FailedRequestException: Local message: failed to read resource at resources/validate: Internal Server Error. Server Message: XDMP-AS: (err:XPTY0004) $uri as xs:string -- Invalid coercion: () as xs:string . See the MarkLogic server error log for further detail.
        at com.marklogic.client.impl.JerseyServices.checkStatus(JerseyServices.java:4610)
        at com.marklogic.client.impl.JerseyServices.getIteratedResourceImpl(JerseyServices.java:3254)
        at com.marklogic.client.impl.JerseyServices.getIteratedResource(JerseyServices.java:3200)
        at com.marklogic.client.impl.ResourceServicesImpl.get(ResourceServicesImpl.java:62)
        at com.marklogic.client.impl.ResourceServicesImpl.get(ResourceServicesImpl.java:58)
        at com.marklogic.hub.DataHub$EntitiesValidator.validate(DataHub.java:344)
        at com.marklogic.hub.DataHub.validateUserModules(DataHub.java:279)
        at com.marklogic.hub.service.DataHubService.validateUserModules(DataHubService.java:44)
        at com.marklogic.hub.web.controller.api.DataHubServerApiController.validateUserModules(DataHubServerApiController.java:173)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)

### PR DESCRIPTION
**Pull request by paxtonhare**
added checks for both types of path separators in regexes

### CODE REVIEW

File: examples/healthcare/plugins/entities/Patients/conformance/final/writer/writer.xqy [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+module namespace plugin = "http://marklogic.com/data-hub/plugins";
+
+declare option xdmp:mapping "false";
+
+(:~
+ : Writer Plugin
+ :
+ : @param $id       - the identifier returned by the collector
+ : @param $envelope - the final envelope
```
Comment (dmcassel) [inline examples/healthcare/plugins/entities/Patients/conformance/final/writer/writer.xqy]: Will the upstream processes always be envelopes? I get that they might be now, but is that the only kind of transform you intend to support? If you have broader capabilities in mind, then maybe

  param $input - the conformed input

File: marklogic-data-hub/src/main/java/com/marklogic/hub/DataHub.java [hunk 2]
```diff
@@ -254,8 +254,8 @@ public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs)
                 boolean isRest = dir.endsWith("REST");
 
                 String dirStr = dir.toString();
-                boolean isInputDir = dirStr.matches(".*/input/.*");
-                boolean isConformanceDir = dirStr.matches(".*/conformance/.*");
+                boolean isInputDir = dirStr.matches(".*[/\\\\]input[/\\\\].*");
+                boolean isConformanceDir = dirStr.matches(".*[/\\\\]conformance[/\\\\].*");
```
Comment (dmcassel) [inline marklogic-data-hub/src/main/java/com/marklogic/hub/DataHub.java]: Are these lines looking at file paths? Should you be using File.separator here?

File: marklogic-data-hub/src/main/java/com/marklogic/hub/util/HubFileFilter.java [hunk 3]
```diff
@@ -10,7 +10,7 @@
 
     @Override
     public boolean accept(File f) {
-        boolean result = f != null && !f.getName().startsWith(".") && !f.toString().matches(".*/REST/.*");
+        boolean result = f != null && !f.getName().startsWith(".") && !f.toString().matches(".*[/\\\\]REST[/\\\\].*");
```
Comment (dmcassel) [inline marklogic-data-hub/src/main/java/com/marklogic/hub/util/HubFileFilter.java]: File.separator?

File: examples/healthcare/plugins/entities/Patients/conformance/final/writer/writer.xqy (hunk 1 above)
Comment (paxtonhare) [inline examples/healthcare/plugins/entities/Patients/conformance/final/writer/writer.xqy]: we have no broader capabilities in mind.

File: marklogic-data-hub/src/main/java/com/marklogic/hub/DataHub.java (hunk 2 above)
Comment (paxtonhare) [inline marklogic-data-hub/src/main/java/com/marklogic/hub/DataHub.java]: You can do either. It's way uglier to do separator, IMO.

```java
boolean isInputDir = dirStr.matches(".*\\" + File.separator + "input\\" + File.separator + ".*");
```

**Pull request merged**